### PR TITLE
Limit order TEAL template

### PR DESCRIPTION
**Pull request by EvanJRichard**
## Summary
This template implements `limit-order-a.teal.impl` from `go-algorand`.

### Testing
Automatic unit/sdk testing. Added a golden test for program and address.

### See also
https://github.com/algorand/js-algorand-sdk/issues/99
https://github.com/algorand/go-algorand-sdk/pull/101

### CODE REVIEW

File: src/logicTemplates/limitorder.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+        let noCloseRemainder = undefined;
+        let noAssetRevocationTarget = undefined;
+        let tx1 = algosdk.makePaymentTxn(this.address, buyerAddr, fee, algoAmount, noCloseRemainder, firstRound, lastRound, undefined, genesisHash, undefined);
+        let tx2 = algosdk.makeAssetTransferTxn(buyerAddr, this.owner, noCloseRemainder, noAssetRevocationTarget, fee, assetAmount, firstRound, lastRound, undefined, genesisHash, undefined, this.assetid);
+
+        let txns = [tx1, tx2];
+        let txGroup = algosdk.assignGroupID(txns);
+
+        let logicSig = algosdk.makeLogicSig(this.getProgram(), undefined); // no args
```
Comment (rotemh) [inline src/logicTemplates/limitorder.js]: I think that this line does the wrong thing. Maybe you've intended to use the `contract`?

File: src/logicTemplates/limitorder.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        let signedTxns = [];
+        for (let idx in txGroup) {
+            let stxn = algosdk.signLogicSigTransaction(txGroup[idx], logicSig);
+            signedTxns.push(stxn)
+        }
```
[the remark this answers is not in the document]
Comment (EvanJRichard) [reply src/logicTemplates/limitorder.js]: hugely good find

File: src/logicTemplates/limitorder.js [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+     */
+    constructor(owner, assetid, ratn, ratd, expiryRound, minTrade, maxFee) {
+        // don't need to validate owner - it will be validated by template.insert
+        if (!Number.isSafeInteger(assetid) || assetid < 0) throw Error("assetid must be a positive number and smaller than 2^53-1");
+        if (!Number.isSafeInteger(ratn) || ratn < 0) throw Error("ratn must be a positive number and smaller than 2^53-1");
+        if (!Number.isSafeInteger(ratd) || ratd < 0) throw Error("ratd must be a positive number and smaller than 2^53-1");
+        if (!Number.isSafeInteger(expiryRound) || expiryRound < 0) throw Error("expiryRound must be a positive number and smaller than 2^53-1");
+        if (!Number.isSafeInteger(minTrade) || minTrade < 0) throw Error("minTrade must be a positive number and smaller than 2^53-1");
+        if (!Number.isSafeInteger(maxFee) || maxFee < 0) throw Error("maxFee must be a positive number and smaller than 2^53-1");
+
```
Comment (btoll) [inline src/logicTemplates/limitorder.js]: Could these validation tests, where I see in other scripts as well, be pulled out into a helper?

File: src/logicTemplates/limitorder.js [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        const referenceProgramB64 = "ASAKAAEFAgYEBwgJCiYBIP68oLsUSlpOp7Q4pGgayA5soQW8tgf8VlMlyVaV9qITMRYiEjEQIxIQMQEkDhAyBCMSQABVMgQlEjEIIQQNEDEJMgMSEDMBECEFEhAzAREhBhIQMwEUKBIQMwETMgMSEDMBEiEHHTUCNQExCCEIHTUENQM0ATQDDUAAJDQBNAMSNAI0BA8QQAAWADEJKBIxAiEJDRAxBzIDEhAxCCISEBA=";
+        let referenceProgramBytes = Buffer.from(referenceProgramB64, 'base64');
```
Comment (pzbitskiy) [inline src/logicTemplates/limitorder.js]: 'base64' -> "base64" (quotes)

File: src/logicTemplates/limitorder.js (hunk 3 above)
Comment (EvanJRichard) [reply src/logicTemplates/limitorder.js]: :+1: Good thinking, will file a ticket

**Pull request merged**